crates/ruff_formatter/src/printer/mod.rs

```diff
@@ -1175,7 +1176,13 @@ impl<'a, 'print> FitsMeasurer<'a, 'print> {
                 }
             }
 
-            FormatElement::Tag(StartLineSuffix) => {
+            FormatElement::Tag(StartLineSuffix(line_suffix)) => {
+                // TODO(cnpryer): Need to understand the relationship between fits checks and prints.
+                //   Not sure if I've ruled out any redundancy or if state should be handled better here.
+                self.state.line_width += line_suffix.reserved_width;
```

Comment:
This seems correct. `fits` runs when trying to see if groups (and the more advanced best fitting and fill elements) fit before printing. For every group (that has mode =Flat, which is true for most), the printer calls fits to see if it fits on the line. If yes, then the printer prints the group in flat mode, otherwise in expanded mode. The fact that `fits` performs a lookahead means that all state changes are reverted when fits end to ensure that the printer continues printing with the same state as when it called into fits.

The printer has a few optimisations to avoid re-measuring the same groups over and over if it already knows that all of them fit anyway. But for reasoning, it's easiest to assume that the printer does this for every group. 

I hope this helps.

Reply:
Helps a lot. Seems like really interesting logic too, so I'll check it out anyway. Thanks!